ethcore/src/state/mod.rs

```diff
@@ -539,8 +542,55 @@ impl<B: Backend> State<B> {
 			|a| a.as_ref().and_then(|account| account.storage_root().cloned()))
 	}
 
-	/// Mutate storage of account `address` so that it is `value` for `key`.
-	pub fn storage_at(&self, address: &Address, key: &H256) -> TrieResult<H256> {
+	/// Get the value of storage at a specific checkpoint.
+	pub fn checkpoint_storage_at(&self, checkpoint_index: usize, address: &Address, key: &H256) -> TrieResult<Option<H256>> {
+		let checkpoints = self.checkpoints.borrow();
+		let checkpoints_len = checkpoints.len();
+		if let Some(ref checkpoint) = checkpoints.get(checkpoint_index) {
+			if let Some(entry) = checkpoint.get(address) {
+				match entry {
+					Some(entry) => {
+						match entry.account {
+							Some(ref account) => {
+								if let Some(value) = account.cached_storage_at(key) {
+									Ok(Some(value))
+								} else {
+									// This key has checkpoint entry, but not in the entry's cache.
+									// So it's not modified at all.
+									Ok(Some(self.original_storage_at(address, key)?))
```

Comment:
Are you sure this is correct? We should probably check all the previous checkpoints, cause maybe the value is there?
I.e:
```
original: Some(Acc { Storage(x) = 0 })
checkpoint0: Some(Acc { Storage(x) = 1 })
checkpoint1: Some(Acc { Storage(x) = None, Storage(y) = 2 })
checkpoint2: Some(Acc { Storage(x) = 3 })
```
so now getting `checkpoint_storage_at(1, _, x)` should most likely return `1`, right? So the state of the account from the previous checkpoint, not the original storage?

Reply:
Can you explain more what you mean by "original" (what's the checkpoint index for "original", "checkpoint0")? If we didn't checkpoint before modifying the state, then the original state is forever lost?

Reply:
You're right, my misunderstanding of how checkpoints work!